src/main/java/io/omnition/loadgenerator/util/SummaryLogger.java

```diff
@@ -0,0 +1,56 @@
+package io.omnition.loadgenerator.util;
+
+import java.util.concurrent.TimeUnit;
+
+import org.apache.log4j.Logger;
+
+/**
+ * SummaryLogger maintains global count of emitted traces and periodically
+ * outputs emitted counters and speed. Intended to be used as a single instance
+ * per app.
+ */
+public class SummaryLogger {
+    // How often to print summary log. 1 second.
+    private static final long SUM_LOG_PERIOD_NANOSEC = TimeUnit.NANOSECONDS.convert(1, TimeUnit.SECONDS);
+
+    // Last time summary log was printed, in monotinic nanoseconds
+    private long lastSumLogTimestampNano;
+
+    // Count of traces emited since summary log last printed
+    private long emitsSinceLastSumLog;
+    private long totalEmits;
+
+    private final Logger logger;
+
+    public SummaryLogger(Logger logger) {
+        this.logger = logger;
+    }
+
+    /**
+     * Count emitted traces and output summary if enough time elapsed since last
+     * output. This function is thread safe and can be called from multipler trace
+     * emitters concurrently.
+     */
+    public synchronized void logEmit(long emittedTraces) {
```

Comment:
perhaps you can even avoid making this synchronized, and update signature slightly:

- onEmitted (traceId) -> this updates its count and perhaps remembers up to 10 sample traceIds
- emit() -> this emits a summary log message (how many traces emitted, some example traceIds)

then emit can be invoked periodically by the caller since anyway it has a scheduled executor

Reply:
Yep, can be done. I would need to use atomic counters to make OnEmitted thread safe. Do you think it is worth spending time on it? My gut feeling is that we are not emitting traces at a pace that will cause measurable contention on this synchronized block (perhaps only if the logger.info blocks) and load generator is not part of out production code to warrant the effort. Happy to change if you think it is worth doing.